lib/preview/previewUtils.js

```diff
@@ -0,0 +1,75 @@
+const { fetchDomains } = require('../../api/domains');
+const { getDirectoryContentsByPath } = require('../../api/fileMapper');
+
+const getPortalDomains = async (portalId) => {
+  try {
+    const result = await fetchDomains(portalId);
+    return result;
+  } catch (error) {
+    console.log("There was a problem fetching domains for your portal. You may be missing a scope necessary for this feature.")
+    return [];
+  }
+}
+
+const getModules = async (sessionInfo) => {
+  const modulesPath = `${sessionInfo.dest}/modules`;
+  return getDirectoryContentsByPath(sessionInfo.portalId, modulesPath)
+    .then(response => response.children)
+    .catch(err => null);
+}
+
+const getTemplates = async (sessionInfo) => {
+  const templatesPath = `${sessionInfo.dest}/templates`;
+  return getDirectoryContentsByPath(sessionInfo.portalId, templatesPath)
+    .then(response => response.children)
+    .catch(err => null);
+}
+
+const getPreviewUrl = (sessionInfo, queryParams) => {
+  const { portalId, env, hublet } = sessionInfo;
+
+  return `http://${portalId}.hubspotpreview${
+    env === 'qa' ? 'qa' : ''
+  }-${hublet}.com/_hcms/preview/template/multi?${stringifyQuery(queryParams)}`;
+}
+
+const stringifyQuery = (query) => {
+  return Object.keys(query)
+    .sort()
+    .map(key => `${key}=${query[key]}`)
+    .join('&');
+}
+
+const insertAtEndOfBody = (html, script) => {
+  const insertAt = (baseStr, index, insertStr) => {
+    return `${baseStr.slice(0, index)}${insertStr}${baseStr.slice(index)}`;
+  }
+  const endOfBodyIndex = html.lastIndexOf("</body>");
+  return insertAt(html, endOfBodyIndex, script);
+}
+
+const addRefreshScript = (html) => {
+  const refreshScript = `
+  <script>
+  setInterval(() => {
+    const res = fetch('http://localhost:3000/meta')
+      .then(async (res) => {
+        const hsServerState = await res.json();
+        if (hsServerState["REMOTE_FS_IS_DIRTY"]) {
+          location.reload();
+        }
+      })
+      .catch(err => console.log(err));
```

Comment:
Could we add a max retry to this? If the user kills the dev server but forgets to close the proxy tab this will fill up the console w/ "failed to fetch" errors. 

It would break a workflow where they kill the process and then restart it without closing the same tab, but not sure if that's a common one, and it's only a refresh away from being fixed.

Reply:
I like an exponential backoff! That keeps the behavior while minimizing the spam